README.md
## Easy 2D Physics in Flutter
This package makes it easy to use a 2D physical engine in flutter.  You can experience acceleration, collision, and rotation.
## Installation
1.  Add this package to your `pubspec.yaml` file.
	```
		dependencies:
			easy_physics_2d: '^0.0.1'
	```
2. Install it.
   ``` 
	   $ pub get
   ```
3. Import.
	```
	    import 'package:easy_physics_2d/easy_physics_2d.dart';
	```
## Usage
```dart
	List<dynamic> objList = [];
	var ball;
	
	@override
	void initState(){
		super.initState();
		ball = myBall(  
					xPoint: 100,  
					yPoint: 200,  
					xVelocity: 0,  
					yVelocity: 0,  
					ballRadius: 30,  
					ballMass: 0.5,  
					angularVelocity: 0,  
				);
		objList = [ball];
	}

	GravityField(  
		objects: objList,  
		mapX: 350,  
		mapY: 350,  
		mapColor: Colors.white  
	),
```
This is the most basic code for using a package. You can add a `GravityField` widget and add a list of physical objects that you want to put in the field.
```dart
	GravityField(  
		objects: objList,  
		gravity: 1000,  
		mapX: 350,  
		mapY: 350,  
		mapColor: Colors.white,
		gravity: 1500,
		frictionConstant: 0.8,
		elasticConstant: 0.9,
	),
```
You can change the detailed properties of the field by changing `gravity`, `frictionConstant` and `elasticConstant`.

For the ball objects, you can also change designs.
```dart
class _HomePageState extends State<HomePage> {  
	Paint paint1 = Paint()  
	  ..color = Color(0xff263e63)  
	  ..style = PaintingStyle.stroke  
	  ..strokeWidth = 2;  

	Paint paint2 = Paint()  
	  ..color = Color(0xff15693b)  
	  ..style = PaintingStyle.stroke  
	  ..strokeWidth = 2;  

	List<Paint> paintList = [];  

	Path draw1 = Path();  
	Path draw2 = Path();  
  
	var ball;  
	var ball2;  
  
	@override  
	void initState() {  
		super.initState();  
		for (double i = 0; i < 20 - 1; i++) {  
			draw1.arcTo(Rect.fromCircle(radius: i, center: Offset(0, 0,)), 0, (1.5 * pi), true);  
			draw2.arcTo(Rect.fromCircle(radius: i, center: Offset(0, 0,)), 1.5 * pi, 0.5 * pi, true);  
		}  

		paintList=[paint1, paint2];  

		ball = myBall(  
		        xPoint: 100,  
				yPoint: 200,  
				xVelocity: 0,  
				yVelocity: 0,  
				ballRadius: 30,  
				ballMass: 0.5,  
				angularVelocity: 0,  
				ballPaint: paintList,
				ballPath: [draw1, draw2], 
		  );  
		ball2 = myBall(  
		        xPoint: 150,  
				yPoint: 100,  
				xVelocity: 0,  
				yVelocity: 0,  
				ballRadius: 20,  
				ballMass: 0.5,  
				angularVelocity: 0,  
				 
		);  
		objList = [ball, ball2];  
	}
}
```
with the parameter `ballPath` and `ballPaint`, you can change `Path`  and Colors of the Ball. At this time, the length of `ballPath` List and the length of `ballPaint` List should be the same, and they should be defined in initState.

Each object has the following instance methods, which are easy to use.
```dart
var ball = myBall(  
	        xPoint: 100,  
			yPoint: 200,  
			xVelocity: 0,  
			yVelocity: 0,  
			ballRadius: 30,  
			ballMass: 0.5,  
			angularVelocity: 0,  
		); 
		 
double n;
double m;
int range1 = 100;
int range2 = 1500;
double x, y;

ball.addXpos(n); //return void
ball.subXpos(n); //return void
ball.addYpos(n); //return void
ball.subYpos(n); //return void

ball.addXvel(m); //return void
ball.subXvel(m); //return void
ball.addYvel(m); //return void
ball.subYvel(m); //return void

ball.stop(); //return void
ball.shuffle(range1, range2); //return void : set velocity randomly range1 to range2

ball.setPosition(x, y); //return void
ball.addAngle(n); //return void

ball.isBallRegion(x, y); //return true if (x, y) is in Ball area

ball.updateDraw(); //return void: if you want tochange the position, you should run this method.
```
If you want to move the `Ball`, please use the instance method to control it.

CHANGELOG.md
## 0.0.1
### 2022.08.30.
* Initial Release


test/easy_physics_2d_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:easy_physics_2d/easy_physics_2d.dart';

void main() {

}


lib/objects.dart
import 'dart:typed_data';
import 'package:flutter/material.dart';
import 'dart:math';


class physicsObject {
  double before_xPos = 0;
  double before_yPos = 0;
  double xPos = 0;
  double yPos = 0;
  double xVel = 0;
  double yVel = 0;
  double xAcc = 0;
  double yAcc = 0;
  double angle = 0;
  double angularVel = 0;
  double angularAcc = 0;

  double mass = 1;
  double rMass = 1;
  double baseTime = 0.016;
  double elasticConstant = 1;
  bool isClick = false;
  bool isClickAfter = true;
  bool isLongClick = false;

  void addXpos(double x) {
    before_xPos = xPos;
    xPos += x;
  }

  void subXpos(double x) {
    before_xPos = xPos;
    xPos -= x;
  }

  void addYpos(double y) {
    before_yPos = yPos;
    yPos += y;
  }

  void subYpos(double y) {
    before_yPos = yPos;
    yPos -= y;
  }

  void addXvel(double x) {
    xVel += x;
  }

  void subXvel(double x) {
    xVel -= x;
  }

  void addYvel(double y) {
    yVel += y;
  }

  void subYvel(double y) {
    yVel -= y;
  }

  void mulXvel(double v) {
    xVel *= v;
  }

  void mulYvel(double v) {
    yVel *= v;
  }

  void stop() {
    xVel = 0;
    yVel = 0;
  }

  void shuffle(int range1, int range2) {
    List vec = [1, -1];
    xVel =
        vec[Random().nextInt(2)] * (Random().nextInt(range2 - range1) + range1);
    yVel =
        vec[Random().nextInt(2)] * (Random().nextInt(range2 - range1) + range1);
  }

  void outVel() {
    if (yVel.abs() < 6.6) {
      yVel = 0;
    }
    if (xVel.abs() < 6.6) {
      xVel = 0;
    }
  }

  void setPosition(double x, double y) {
    before_xPos = xPos;
    before_yPos = yPos;
    xPos = x;
    yPos = y;
  }

  void addAngle(double ang) {
    angle += ang;
  }
}

class myBall extends physicsObject {
  late double ballRad;
  List<Path> draws = [];
  List<Path> backup_draws = [];
  List<Paint> paints = [];

  String objType = 'ball';
  late double momentI;

  double xPoint;
  double yPoint;
  double xVelocity;
  double yVelocity;
  double ballRadius;
  double ballMass;
  double angularVelocity;
  List<Path> ballPath = [];
  List<Paint> ballPaint = [];

  myBall(
      {Key? key,
        required this.xPoint,
        required this.yPoint,
        required this.xVelocity,
        required this.yVelocity,
        required this.ballRadius,
        required this.ballMass,
        required this.angularVelocity,
        this.ballPath = const [],
        this.ballPaint = const [],
      }) {
    before_xPos = xPos;
    before_yPos = yPos;
    super.xPos = xPoint;
    super.yPos = yPoint;
    super.xVel = xVelocity;
    super.yVel = yVelocity;
    super.mass = ballMass;
    super.rMass = 1 / ballMass;
    ballRad = ballRadius;

    if(ballPath.length == 0){
      Path draw1 = Path();
      Path draw2 = Path();
      for (double i = 0; i < ballRad - 1; i++) {
        draw1.arcTo(
            Rect.fromCircle(
              radius: i,
              center: Offset(
                0,
                0,
              ),
            ),
            0 ,
            (1.9 * pi),
            true);

        draw2.arcTo(
            Rect.fromCircle(
              radius: i,
              center: Offset(
                0,
                0,
              ),
            ),
            1.9 * pi ,
            0.1 * pi,
            true);
      }


      draws.add(draw1);
      draws.add(draw2);

      backup_draws.add(draw1);
      backup_draws.add(draw2);
    }
    else{
      draws=ballPath;
      backup_draws=ballPath.toList();
    }


    if(ballPaint.length == 0) {
      Paint paint1 = Paint()
        ..color = Color(0xffcce0ff)
        ..style = PaintingStyle.stroke
        ..strokeWidth = 2;

      Paint paint2 = Paint()
        ..color = Color(0xffb0cfff)
        ..style = PaintingStyle.stroke
        ..strokeWidth = 2;

      paints.add(paint1);
      paints.add(paint2);
    }
    else{
      paints=ballPaint;
    }

    angularVel = angularVelocity;
    momentI = 0.5 * mass * ballRad * ballRad;


  }

  bool isBallRegion(double checkX, double checkY) {
    if ((pow(super.xPos - checkX, 2) + pow(super.yPos - checkY, 2)) <=
        pow(ballRad, 2)) {
      return true;
    }
    return false;
  }


  void updateDraw() {
    var translationMatrix = Float64List.fromList([
      cos(angle), sin(angle), 0, 0,
      -sin(angle), cos(angle), 0, 0,
      0, 0, 1, 0,
      0, 0, 0, 1
    ]);

    for(int i=0; i<draws.length; i++){
      draws[i] = backup_draws[i].transform(translationMatrix);
      draws[i] = draws[i].shift(Offset(xPos, yPos));
    }

  }
}


lib/gravity_field.dart
import 'dart:async';
import 'dart:math';
import 'objects.dart';
import 'package:flutter/material.dart';

class GravityField extends StatefulWidget {
  final List<dynamic> objects;
  final double gravity;
  final double mapX;
  final double mapY;
  final Color mapColor;
  final double frictionConstant;
  final double elasticConstant;

  const GravityField({
    Key? key,
    required this.objects,
    this.gravity = 600,
    required this.mapX,
    required this.mapY,
    required this.mapColor,
    this.frictionConstant = 0.7,
    this.elasticConstant = 0.8
  }) : super(key: key);

  @override
  _GravityFieldState createState() => _GravityFieldState();
}

class _GravityFieldState extends State<GravityField> with SingleTickerProviderStateMixin {
  bool isClick = false;
  bool isClickAfter = true;
  double mapY = 0;
  double mapX = 0;
  double elasticConstant = 0.8;
  List objList = [];
  List objPathList = [];
  List objPaintList = [];
  Color map_color = Colors.white;

  List pathList = [];

  late AnimationController _animationController;
  double baseTime = 0.016;
  int milliBaseTime = 16;
  double gravityAccel = 600;
  double frictionC = 0.7;

  List iPos = [];
  List fPos = [];

  double timerMilllisecond = 0;
  int longclickobj = 0;

  bool collapse = false;

  @override
  void initState() {
    // TODO: implement initState
    super.initState();

    _animationController =
        AnimationController(vsync: this, duration: Duration(milliseconds: 1));
    _animationController.repeat();

    objList = widget.objects;
    gravityAccel = widget.gravity;
    mapX = widget.mapX;
    mapY = widget.mapY;
    map_color = widget.mapColor;
    elasticConstant = widget.elasticConstant;
    frictionC = widget.frictionConstant;

    objList.forEach((e) {
      objPaintList.add(e.paints);
      objPathList.add(e.draws);
      e.yAcc=gravityAccel;
    });

  }

  @override
  void dispose() {
    _animationController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onVerticalDragDown: (details) {
        for (var i = 0; i < objList.length; i++) {
          setState(() {
            if (objList[i].isBallRegion(
                details.localPosition.dx, details.localPosition.dy)) {
              objList[i].isClick = true;

              objList[i].stop();

            }
          });
        }
      },
      onVerticalDragEnd: (details) {
        for (var i = 0; i < objList.length; i++) {
          if (objList[i].isClick) {
            setState(() {
              objList[i].isClick = false;
              objList[i].isClickAfter = true;
            });
          }
        }
      },
      onLongPressDown: (details) {
        for (var i = 0; i < objList.length; i++) {
          setState(() {
            if (objList[i].isBallRegion(
                details.localPosition.dx, details.localPosition.dy)) {
              iPos.add(details.localPosition.dx);
              iPos.add(details.localPosition.dy);
              objList[i].isLongClick = true;
              longclickobj = i;
            }
          });
        }
      },
      onLongPressEnd: (details) {
        if (objList[longclickobj].isLongClick) {
          setState(() {
            objList[longclickobj].xVel =
                3 * (details.localPosition.dx - iPos[0]) / (0.7);
            objList[longclickobj].yVel =
                3 * (details.localPosition.dy - iPos[1]) / (0.7);

            objList[longclickobj].isLongClick = false;
            objList[longclickobj].isClick = false;
            objList[longclickobj].isClickAfter = true;
          });
        }

        iPos = [];
        fPos = [];
      },
      onVerticalDragUpdate: (details) {
        for (var i = 0; i < objList.length; i++) {
          if (objList[i].isClick) {
            setState(() {
              objList[i].setPosition(
                  details.localPosition.dx, details.localPosition.dy);
              objList[i].updateDraw();
            });
          }
        }
      },
      child: AnimatedBuilder(
          animation: _animationController,
          builder: (context, child) {
            int wallInt = 0;
            bool Collapse = false;
            bool flag = false;

            for (var i = 0; i < objList.length; i++) {
              if (!objList[i].isClick) {
                objList[i].addYvel(baseTime * objList[i].yAcc);
                objList[i].addXvel(baseTime * objList[i].xAcc);

                checkCollapse(objList, baseTime);

                wallInt = checkIsWall(objList[i]);

                if (checkIsWall(objList[i]) == 1) {

                  flag = true;

                  double wallCorrection = objList[i].yVel * baseTime +
                      objList[i].yPos +
                      objList[i].ballRad -
                      mapY -
                      1;
                  objList[i].subYpos(wallCorrection);


                  double vxi = objList[i].xVel;
                  double vyi = objList[i].yVel;

                  double xi = objList[i].xPos;
                  double yi = objList[i].yPos;

                  List<double> iPos = [xi, yi];

                  List<double> pPos = [xi, mapY];
                  List<double> ripVec = [pPos[0] - iPos[0], pPos[1] - iPos[1]];
                  List<double> ripNorVec = [
                    ripVec[0] /
                        sqrt(ripVec[0] * ripVec[0] + ripVec[1] * ripVec[1]),
                    ripVec[1] /
                        sqrt(ripVec[0] * ripVec[0] + ripVec[1] * ripVec[1])
                  ];

                  List<double> vipVec = [vxi, vyi];
                  List<double> vjpVec = [0, 0];
                  List<double> vijVec = [
                    vipVec[0] - vjpVec[0],
                    vipVec[1] - vjpVec[1]
                  ];


                  double pulse = (-(1 + elasticConstant) *
                      innerProduct(vijVec, ripNorVec)) /
                      ((objList[i].rMass) * innerProduct(ripNorVec, ripNorVec) +
                          (innerProduct(ripVec, ripNorVec) *
                              innerProduct(ripVec, ripNorVec)) /
                              objList[i].momentI);

                  double wi = objList[i].angularVel;


                  objList[i].mulYvel(-elasticConstant);
                  objList[i].mulXvel(elasticConstant);
                  objList[i].angularVel *= frictionC;
                }
                if (checkIsWall(objList[i]) == 2) {
                  flag = true;
                  objList[i].mulYvel(-elasticConstant);
                  objList[i].mulXvel(elasticConstant);
                  objList[i].angularVel *= frictionC;
                }
                if (checkIsWall(objList[i]) == 3) {
                  flag = true;
                  double wallCorrection =
                      objList[i].xPos + objList[i].ballRad - mapX + 1;
                  objList[i].subXpos(wallCorrection);
                  objList[i].mulXvel(-elasticConstant);
                  objList[i].mulYvel(elasticConstant);
                  objList[i].angularVel *= frictionC;
                }
                if (checkIsWall(objList[i]) == 4) {
                  flag = true;
                  double wallCorrection =
                      -objList[i].xPos + objList[i].ballRad + 1;
                  objList[i].addXpos(wallCorrection);
                  objList[i].mulXvel(-elasticConstant);
                  objList[i].mulYvel(elasticConstant);
                  objList[i].angularVel *= frictionC;
                }
                if (!flag && (checkIsWall(objList[i]) == 0)) {
                  //print(objList[i].yVel);
                  objList[i].addYpos(objList[i].yVel * baseTime);
                  objList[i].addXpos(objList[i].xVel * baseTime);
                }

                objList[i].angularVel += baseTime * objList[i].angularAcc;
                objList[i].addAngle(baseTime * objList[i].angularVel);

                flag = false;
              }

              objList[i].updateDraw();
            }


            return Container(
              width: mapX,
              height: mapY,
              color: map_color,
              child: CustomPaint(
                //painter: _paint(pathList: [ball.draw, newball.draw]),
                painter: _paint(
                  //pathList: [objList[0].draws, objList[1].draws] objPathList,
                  pathList: objPathList,
                  paintList: objPaintList,
                  //paintList: [objList[0].paints, objList[1].paints] objPaintList,
                ),
              ),
            );
          }),
    );
  }

  void checkCollapse(List<dynamic> objList, double baseTime) {
    for (int i = 0; i < objList.length; i++) {
      for (int j = i + 1; j < objList.length; j++) {
        if (getDistance(objList[i], objList[j]) <
            (objList[i].ballRad + objList[j].ballRad)) {
          double correctDistance = 0.5 *
              ((objList[i].ballRad + objList[j].ballRad) -
                  getDistance(objList[i], objList[j]) +
                  2);

          double eConstant = elasticConstant;

          double vxi = objList[i].xVel;
          double vyi = objList[i].yVel;
          double vxj = objList[j].xVel;
          double vyj = objList[j].yVel;

          double xi = objList[i].xPos;
          double yi = objList[i].yPos;
          double xj = objList[j].xPos;
          double yj = objList[j].yPos;

          double ri = objList[i].ballRad;
          double rj = objList[j].ballRad;

          List<double> iPos = [xi, yi];
          List<double> jPos = [xj, yj];

          List<double> pPos = [
            (ri * xj + rj * xi) / (ri + rj),
            (ri * yj + rj * yi) / (ri + rj)
          ];
          List<double> ripVec = [pPos[0] - iPos[0], pPos[1] - iPos[1]];
          List<double> ripNorVec = [
            ripVec[0] / sqrt(ripVec[0] * ripVec[0] + ripVec[1] * ripVec[1]),
            ripVec[1] / sqrt(ripVec[0] * ripVec[0] + ripVec[1] * ripVec[1])
          ];
          List<double> rjpVec = [pPos[0] - jPos[0], pPos[1] - jPos[1]];
          List<double> rjpNorVec = [
            rjpVec[0] / sqrt(rjpVec[0] * rjpVec[0] + rjpVec[1] * rjpVec[1]),
            rjpVec[1] / sqrt(rjpVec[0] * rjpVec[0] + rjpVec[1] * rjpVec[1])
          ];

          List<double> vipVec = [vxi, vyi];
          List<double> vjpVec = [vxj, vyj];
          List<double> vijVec = [vipVec[0] - vjpVec[0], vipVec[1] - vjpVec[1]];


          double viScala = sqrt(vxi * vxi + vyi * vyi);
          double vjScala = sqrt(vxj * vxj + vyj * vyj);


          objList[i].subXpos((viScala / (viScala + vjScala)) * correctDistance * ripNorVec[0]);
          objList[i].subYpos((viScala / (viScala + vjScala)) * correctDistance * ripNorVec[1]);


          objList[j].subXpos((vjScala / (viScala + vjScala)) * correctDistance * rjpNorVec[0]);
          objList[j].subYpos((vjScala / (viScala + vjScala)) * correctDistance * rjpNorVec[1]);


          double pulse = (-(1 + eConstant) * innerProduct(vijVec, ripNorVec)) /
              ((objList[i].rMass + objList[j].rMass) *
                  innerProduct(ripNorVec, ripNorVec) +
                  (innerProduct(ripVec, rjpNorVec) *
                      innerProduct(ripVec, rjpNorVec)) /
                      objList[i].momentI +
                  (innerProduct(rjpVec, rjpNorVec) *
                      innerProduct(rjpVec, rjpNorVec)) /
                      objList[j].momentI);

          double wi = objList[i].angularVel;
          double wj = objList[j].angularVel;
          //
          objList[i].xVel = vipVec[0] + pulse * objList[i].rMass * ripNorVec[0];
          objList[i].yVel = vipVec[1] + pulse * objList[i].rMass * ripNorVec[1];
          objList[j].xVel = vjpVec[0] + pulse * objList[j].rMass * rjpNorVec[0];
          objList[j].yVel = vjpVec[1] + pulse * objList[j].rMass * rjpNorVec[1];

          objList[i].angularVel = wi +
              pulse *
                  innerProduct([ripVec[1], ripVec[0]], ripNorVec) /
                  (objList[i].momentI);
          objList[j].angularVel = wj -
              pulse *
                  innerProduct([rjpVec[1], rjpVec[0]], ripNorVec) /
                  (objList[j].momentI);
        }
      }
    }
  }

  int checkIsWall(var obj) {
    if (obj.yVel * baseTime + obj.yPos + obj.ballRad >= mapY) {
      return 1;
    } else if (obj.yVel * baseTime + obj.yPos - obj.ballRad <= 0) {
      return 2;
    } else if (obj.xVel * baseTime + obj.xPos + obj.ballRad >= mapX) {
      return 3;
    } else if (obj.xVel * baseTime + obj.xPos - obj.ballRad <= 0) {
      return 4;
    }

    return 0;
  }
}

double getL2norm(List vec) {
  return vec[0] * vec[0] + vec[1] * vec[1];
}

double innerProduct(List vec1, List vec2) {
  return vec1[0] * vec2[0] + vec1[1] * vec2[1];
}

double getDistance(physicsObject obj1, physicsObject obj2) {
  return sqrt((obj1.xPos - obj2.xPos) * (obj1.xPos - obj2.xPos) +
      (obj1.yPos - obj2.yPos) * (obj1.yPos - obj2.yPos));
}

class _paint extends CustomPainter {
  final List pathList;
  final List paintList;

  _paint({
    required this.pathList,
    required this.paintList,
  });

  @override
  void paint(Canvas canvas, Size size) {
    Path path = Path();

    for (var i = 0; i < pathList.length; i++) {
      //canvas.drawShadow(pathList[i][0], Colors.grey, sqrt(10), false);
      //path.addPath(pathList[i], Offset.zero);
      canvas.drawPath(pathList[i][0], paintList[i][0]);
      canvas.drawPath(pathList[i][1], paintList[i][1]);
    }
    //canvas.drawPath(path, paintList[i][0]);
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) => true;
}


example/lib/main.dart
import 'dart:math';
import 'package:easy_physics_2d/gravity_field.dart';
import 'package:easy_physics_2d/objects.dart';
import 'package:flutter/material.dart';



void main() {
  runApp(const MyApp());
}

class colorLibrary {
  static Color mainColor2 = const Color.fromARGB(255, 255, 248, 235);
  static Color mapColor = const Color(0xffefe0c3);
  static Color mainColor = const Color.fromARGB(255, 214, 237, 255);
  static Color buttonColor = const Color(0xffffd6a9);
}


class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    final newTextTheme = Theme.of(context).textTheme.apply(
      bodyColor: Colors.white70,
      displayColor: Colors.black87,
    );

    return MaterialApp(
        title: 'Flutter Physics World',
        debugShowCheckedModeBanner: false,
        theme: ThemeData(
            scaffoldBackgroundColor: colorLibrary.mainColor2,
            //primarySwatch: Colors.indigo,
            textTheme: newTextTheme),
        home: HomePage());
  }
}


class HomePage extends StatefulWidget {
  const HomePage({Key? key}) : super(key: key);

  @override
  _HomePageState createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  double sliderValue = 1000;

  Paint paint1 = Paint()
    ..color = Color(0xff263e63)
    ..style = PaintingStyle.stroke
    ..strokeWidth = 2;

  Paint paint2 = Paint()
    ..color = Color(0xff15693b)
    ..style = PaintingStyle.stroke
    ..strokeWidth = 2;

  List<Paint> paintList = [];
  List<dynamic> objList = [];

  Path draw1 = Path();
  Path draw2 = Path();

  var ball;
  var ball2;


  @override
  void initState() {
    super.initState();

    for (double i = 0; i < 20 - 1; i++) {
      draw1.arcTo(
          Rect.fromCircle(
            radius: i,
            center: Offset(
              0,
              0,
            ),
          ),
          0 ,
          (1.5 * pi),
          true);

      draw2.arcTo(
          Rect.fromCircle(
            radius: i,
            center: Offset(
              0,
              0,
            ),
          ),
          1.5 * pi ,
          0.5 * pi,
          true);
    }

    paintList=[paint1, paint2];
    ball = myBall(
        xPoint: 100,
        yPoint: 200,
        xVelocity: 0,
        yVelocity: 0,
        ballRadius: 30,
        ballMass: 0.5,
        angularVelocity: 0,
        ballPaint: paintList
    );
    ball2 = myBall(
      xPoint: 150,
      yPoint: 100,
      xVelocity: 0,
      yVelocity: 0,
      ballRadius: 20,
      ballMass: 0.5,
      angularVelocity: 0,
      ballPath: [draw1, draw2],
    );
    objList = [ball, ball2];
  }

  @override
  Widget build(BuildContext context) {
    //print(objList.length);

    return Scaffold(
      body: _buldBody(),
    );
  }

  Widget _buldBody() {
    //print(ball.xPos);



    return Container(
      child: Column(
        mainAxisAlignment: MainAxisAlignment.spaceEvenly,
        children: [
          Text(
            "Flutter Physics",
            style: TextStyle(
                fontWeight: FontWeight.w800,
                fontSize: 40,
                color: Colors.black45),
          ),

          Container(
            child: GravityField(
              objects: objList,
              gravity: sliderValue,
              mapX: 350,
              mapY: 350,
              mapColor: colorLibrary.mapColor,

            ),
            padding: EdgeInsets.all(1),
          ),

          Row(mainAxisAlignment: MainAxisAlignment.spaceAround, children: [
            Container(
              height: 60,
              width: 60,
              decoration: BoxDecoration(
                  borderRadius: BorderRadius.circular(10),
                  color: colorLibrary.buttonColor),
              child: IconButton(
                icon: Icon(Icons.stop_circle_outlined,
                    color: Colors.white, size: 40),
                onPressed: () {
                  for (int i = 0; i < objList.length; i++) {
                    objList[i].stop();
                  }
                },
              ),
            ),
            Container(
              height: 60,
              width: 60,
              decoration: BoxDecoration(
                  borderRadius: BorderRadius.circular(10),
                  color: colorLibrary.buttonColor),
              child: IconButton(
                icon: Icon(Icons.shuffle, color: Colors.white, size: 40),
                onPressed: () {
                  for (int i = 0; i < objList.length; i++) {
                    objList[i].shuffle(100, 1500);
                  }
                },
              ),
            ),

            Row(
              children: [
                Text('Gravity :', style: TextStyle(fontSize: 20, color: Colors.brown),),
                SliderTheme(
                  data: SliderThemeData(
                    activeTrackColor: Colors.brown,
                    thumbColor: Colors.brown,
                    activeTickMarkColor: Colors.brown,
                    valueIndicatorColor: Colors.brown,
                    valueIndicatorShape: PaddleSliderValueIndicatorShape(),
                  ),
                  child: Slider(
                    value: sliderValue,
                    min: 0.0,
                    max: 3000.0,
                    divisions: 50,
                    label: '${sliderValue.toInt()}',
                    onChanged: (double newValue) {
                      setState( () {
                        sliderValue = newValue;
                        ball.yAcc = newValue;
                        ball2.yAcc = newValue;
                      },
                      );
                    },
                  ),
                ),
              ],
            ),
          ])
        ],
      ),
    );
  }
}
